openqa-bats-review

```diff
@@ -0,0 +1,208 @@
+
+
+TIMEOUT = 30
+USER_AGENT = "openqa-bats-review (https://github.com/os-autoinst/scripts)"
+
+logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
+log = logging.getLogger(sys.argv[0] if __name__ == "__main__" else __name__)
+session = requests.Session()
+
+
+# We want to extract the test name stripping the numbers and the optional timing information
+# not ok 166 bud-git-context in 118ms
+# not ok 655 [520] podman checkpoint --export, with volumes in 1558ms
+NOT_OK = re.compile(r"^not ok \d+ (?:\[\d+\] )?(.*)(?: in \d+ms)?$")
+
+
+def get_file(url: str) -> str:
+    """
+    Get a text file from URL
+    """
+    headers = {
+        "User-Agent": USER_AGENT,
+    }
+    try:
+        got = session.get(url, headers=headers, timeout=TIMEOUT)
+        got.raise_for_status()
+    except RequestException as error:
+        log.error("%s: %s", url, error)
+        sys.exit(1)
+    return got.text
+
+
+@cache
+def get_job(url: str) -> dict:
+    """
+    Get a job from openQA
+    """
+    headers = {
+        "User-Agent": USER_AGENT,
+    }
+    try:
+        got = session.get(url, headers=headers, timeout=TIMEOUT)
+        got.raise_for_status()
+        data = got.json()
+    except RequestException as error:
+        log.error("%s: %s", url, error)
+        sys.exit(1)
+    return data["job"]
+
+
+def grep_notok(url: str) -> set[str]:
+    """
+    Grep for "not ok" lines and return a set with the failing tests
+    prefixed by the filename
+    """
+    notok = set()
+    # Note: We use a prefix because some tests have more than 1 TAP file
+    prefix = os.path.basename(url)
+    data = get_file(url)
+    lines = data.splitlines()
+    # Fetch the plan to check that we don't have a truncated TAP file
+    last = 0
+    for line in lines:
+        if line.startswith("1.."):
+            last = int(line.split("..", 1)[1])
+            break
+    if not last:
+        log.error("Malformed TAP file: %s", url)
+        sys.exit(1)
+    # Trust the plan
+    tests = 0
+    for line in lines:
+        if not line.startswith(("ok", "not ok", "#not ok")):
+            continue
+        tests += 1
+        try:
+            test = NOT_OK.findall(line)[0]
+            notok.add(f"{prefix}:{test}")
+        except IndexError:
+            continue
+    if tests != last:
+        log.error("Truncated TAP file: %s", url)
+        sys.exit(1)
+    return notok
+
+
+def process_files(files: list[str]) -> set[str]:
+    """
+    Process TAP files
+    """
+    # The test for these packages have only one TAP file: aardvark-dns & netavark
+    if len(files) == 1:
+        return grep_notok(files[0])
+    # Use multithreading for tests like buildah, runc & skopeo (2 files) & podman (4 files)
+    with ThreadPoolExecutor(max_workers=len(files)) as executor:
+        return set(itertools.chain.from_iterable(executor.map(grep_notok, files)))
+
+
+def get_clone_chain(openqa_host: str, job_id: int) -> list[int]:
```

Comment:
To convey at least a little bit better that this is an expensive operation:
```suggestion
def resolve_clone_chain(openqa_host: str, job_id: int) -> list[int]:
```

Note that openQA itself provides in its internal code functions to resolve this much more efficiently. If needed/wanted we could expose this via an API route and avoid duplicating the logic here and a worse way. (Perhaps this even is already exposed via *some* route. I needed to check this myself, too.)

Reply:
We have to fetch the logs anyway so I don't see the problem with this approach as the results are cached.

Reply:
Yes, I mean ancestors/descendants. It would of course be easy to add a route (or add parameters to an existing route to compute/return ancestors/descendants as well).

Reply:
Please note that creating a new API route is overkill given that I need to fetch `/details` anyway so I can get the log filenames.